clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -127,6 +128,24 @@ public void awaitMetadataUpdate() {
     }
 
     /**
+     * Block until the metadata has been refreshed or the timeout has expired.
+     * @param timeout The maximum duration (in ms) to wait for the request
+     * @throws TimeoutException if the timeout has expired
+     */
+    public void awaitMetadataUpdate(long timeout) {
+        long now = time.milliseconds();
+        long deadline = now + timeout;
+        int version = this.metadata.requestUpdate();
+        while (this.metadata.version() == version && now < deadline) {
```

Comment:
Thanks for the quick reply @ijuma. Yeah the point you raise is valid.

It looks like KafkaConsumer.listTopics (more specifically, its call to Fetcher.getAllTopics) in this RB is actually doing 1, but KafkaConsumer.partitionsFor (more specifically, its call to ConsumerNetworkClient.awaitMetadataUpdate) is doing 3. Maybe we should make them both act the same.

Also just a heads up, this PR is very much WIP. It only has listTopics and partitionsFor with timeouts so far. I put this PR up partially so that it shows the flaw in using request.timeout.ms as the timeout. From the jira ticket:
- if request.timeout.ms < session.timeout.ms, then a user may drop JoinGroupResponses which can arrive worst case after session.timeout.ms.
- if request.timeout.ms > session.timeout.ms, then a user calling one of these blocking calls will lose their session and trigger a rebalance.